geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaProducerFeatureStore.scala

```diff
@@ -64,86 +64,83 @@ class KafkaProducerFeatureStore(entry: ContentEntry,
     case _              => super.removeFeatures(filter)
   }
 
-  def clearFeatures(): Unit = {
-    val msg = GeoMessage.clear()
-    logger.debug("sending message: {}", msg)
+  def clearFeatures(): Unit = writerPool.withResource { fw => fw.send(GeoMessage.clear()) }
 
-    val encoder = new KafkaGeoMessageEncoder(sft)
-    producer.send(encoder.encodeClearMessage(topic, msg))
+  override def getWriterInternal(query: Query, flags: Int) = {
+    if (query == null || query == Query.ALL || query.getFilter == null || query.getFilter == Filter.INCLUDE) {
+      new KafkaFeatureWriterAppend(sft, producer, topic)
+    } else {
+      new KafkaFeatureWriterModify(sft, producer, topic, query)
+    }
   }
 
-  override def getWriterInternal(query: Query, flags: Int) =
-    new ModifyingFeatureWriter(query)
+  override def getCountInternal(query: Query): Int = 0
+  override def getReaderInternal(query: Query): FeatureReader[SimpleFeatureType, SimpleFeature] = null
 
-  class ModifyingFeatureWriter(query: Query) extends FeatureWriter[SimpleFeatureType, SimpleFeature] with LazyLogging {
+  override def close(): Unit = producer.close()
+}
 
-    val msgEncoder = new KafkaGeoMessageEncoder(sft)
-    val reuse = new ScalaSimpleFeature("", sft)
+abstract class KafkaFeatureWriter(sft: SimpleFeatureType, producer: Producer[Array[Byte], Array[Byte]], topic: String)
+    extends FeatureWriter[SimpleFeatureType, SimpleFeature]with LazyLogging {
 
-    private var id = 1L
-    def getNextId: String = {
-      val ret = id
-      id += 1
-      s"$ret"
-    }
+  protected val msgEncoder = new KafkaGeoMessageEncoder(sft)
 
-    var toModify: Iterator[SimpleFeature] =
-      if(query == null) Iterator[SimpleFeature]()
-      else if(query.getFilter == null) Iterator.continually {
-        reuse.getIdentifier.setID(getNextId)
-        reuse
-      }
-      else query.getFilter match {
-        case ids: Id        =>
-          ids.getIDs.map(id => new ScalaSimpleFeature(id.toString, sft)).iterator
+  private [kafka] def write(sf: SimpleFeature): Unit = send(GeoMessage.createOrUpdate(sf))
 
-        case Filter.INCLUDE =>
-          Iterator.continually(new ScalaSimpleFeature("", sft))
-      }
+  private [kafka] def send(msg: GeoMessage): Unit = {
+    logger.debug("sending message: {}", msg)
+    producer.send(msgEncoder.encodeMessage(topic, msg))
+  }
 
-    def setIter(iter: Iterator[SimpleFeature]): Unit = {
-      toModify = iter
-    }
+  override def getFeatureType: SimpleFeatureType = sft
+  override def close(): Unit = {}
+}
 
-    var curFeature: SimpleFeature = null
-    override def getFeatureType: SimpleFeatureType = sft
+class KafkaFeatureWriterAppend(sft: SimpleFeatureType, producer: Producer[Array[Byte], Array[Byte]], topic: String)
```

Comment:
Yeah I'm not really sure...but I assumed it that ```getFeaturerWriterAppend``` was mostly an optimization because it wouldn't need to update anything features or read features it could simply write them which for some datastores would be a different access pattern or access different parts of the underlying store's code